samples/shell/src/golioth.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (c) 2022 Golioth, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_DECLARE(golioth_shell);
+
+#include <logging/golioth.h>
+#include <net/golioth.h>
+#include <stdlib.h>
+#include <zephyr/net/coap.h>
+#include <zephyr/shell/shell.h>
+
+#include "common.h"
+
+static struct golioth_client golioth_shell_client;
+struct golioth_client *client = &golioth_shell_client;
+
+static int cmd_golioth_connect(const struct shell *sh, size_t argc, char **argv)
+{
+	const char *host = argv[1];
+	unsigned long port = 5684;
+	int err;
+
+	if (argc > 2) {
+		const char *port_s = argv[2];
+
+		port = strtoul(port_s, NULL, 0);
+	}
+
+	err = golioth_connect(client, host, port);
+	if (err) {
+		if (err == -EALREADY) {
+			shell_warn(sh, "Already connected");
+		} else {
+			shell_error(sh, "Failed to connect: %d", err);
+		}
+
+		return err;
+	}
+
+	return 0;
+}
+
+static int cmd_golioth_disconnect(const struct shell *sh, size_t argc, char **argv)
+{
+	int err;
+
+	err = golioth_disconnect(client);
+	if (err) {
+		if (err == -ENOTCONN) {
+			shell_warn(sh, "Already disconnected");
+		} else {
+			shell_error(sh, "Failed to disconnect: %d", err);
+		}
+
+		return err;
+	}
+
+	return 0;
+}
+
+static int cmd_golioth_is_connected(const struct shell *sh, size_t argc, char **argv)
+{
+	bool connected;
+
+	connected = golioth_is_connected(client);
+	shell_print(sh, "%s", connected ? "Connected" : "Disconnected");
+
+	return 0;
+}
+
+static int cmd_golioth_set_coap_dtls(const struct shell *sh, size_t argc, char **argv)
+{
+	static sec_tag_t *tags;
+	static size_t num_tags;
+	int err;
+
+	free(tags);
```

Comment:
>it will be initialized to zero, so I think this would effectively be free(0) the first time it executes

True!

From `man 3 free`:

> The free() function frees the memory space pointed to by ptr, which must have been returned by a previous call to malloc(), calloc(), or realloc().
Otherwise, or if free(ptr) has already been called before, undefined behavior occurs.  **If ptr is NULL, no operation is performed.**

Assigning `NULL` to pointer is one of the practices I've seen to prevent undefined behavior of calling `free()` twice on the same pointer. This might be just a workaround lots of time for unbalanced malloc/free, but in this case I used this fact on purpose to reduce number of if-else paths (branches) in the code and hence simplification of the code.

Reply:
Okay, I see. I didn't know that about `free`. It's still a little odd (to me) to see something freed before it's ever allocated. Perhaps a comment above would be helpful?